tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c

```diff
@@ -0,0 +1,205 @@
+ * in any user struct
+ *
+ * @details
+ * Initialize a user defined struct contains rbtree node
+ * Add into a rbtree
+ * Enumerate the rbtree node
+ *
+ * @ingroup lib_rbtree_tests
+ *
+ * @see RB_FOR_EACH(),RB_FOR_EACH_CONTAINER()
+ */
+void test_rbtree_container(void)
+{
+	int count = 0;
+	static struct rbtree test_tree_l;
+	struct container_node tree_node[10];
+	struct container_node *c_foreach_node;
+	struct rbnode *foreach_node;
+
+	memset(&test_tree_l, 0, sizeof(test_tree_l));
+	test_tree_l.lessthan_fn = node_lessthan;
+	memset(tree_node, 0, sizeof(tree_node));
+	for (unsigned int i = 0; i < 10; i++) {
+		tree_node[i].value = i;
+		rb_insert(&test_tree_l, &tree_node[i].node);
+	}
+
+	/*"for each" style iteration to verify node*/
+	RB_FOR_EACH(&test_tree_l, foreach_node) {
+		zassert_true(((struct container_node *)foreach_node)->value
+				== count, "RB_FOR_EACH failed");
+		count++;
+	}
+
+	count = 0;
+	/*"for each" style iteration to verify node*/
+	RB_FOR_EACH_CONTAINER(&test_tree_l, c_foreach_node, node) {
+		zassert_true(c_foreach_node->value == count,
+				"RB_FOR_EACH_CONTAINER failed");
+		count++;
+	}
+}
+
+static unsigned int next_rand_mod(unsigned int mod)
+{
+	static unsigned long long state = 123456789; /* seed */
+
+	state = state * 2862933555777941757ul + 3037000493ul;
+
+	return ((unsigned int)(state >> 32)) % mod;
+}
+
+static struct rbnode nodes[512];
+static int random_num[512] = {0};
+void random_generate(int size)
+{
+	unsigned int ret = 0;
+	int detect = 0;
+
+	for (int i = 0; i < 512; i++) {
+		ret = next_rand_mod(size);
+		for (int j = 0; j < i; j++) {
+			if (ret == random_num[j]) {
+				detect = 1;
+				break;
+			}
+		}
+		/* writing rand numbers into array,
+		 * if the random number and the existing data duplicate,
+		 * skip this bit, keep 0.
+		 */
+		if (detect != 1) {
+			random_num[i] = ret;
+		}
+		detect = 0;
+	}
+}
+
+static struct rbtree tree;
+int init_tree(struct rbtree *tree, int size)
+{
+	int nodes_count = 0;
+	int temp = 0;
+
+	tree->lessthan_fn = node_lessthan;
+	random_generate(size);
+	for (int i = 0; i < size; i++) {
+		/*skip the bit whitch is 0 to avoid inserting repeatedly*/
+		if (random_num[i]) {
+			temp = random_num[i];
+			rb_insert(tree, &nodes[temp]);
+			nodes_count++;
+		}
+	}
+
+	return nodes_count;
+}
+
+int search_get_minmax_times(struct rbtree *tree, int size)
+{
+	int operations_count = 0;
+	struct rbnode *rx_node = tree->root;
+	struct rbnode *temp = NULL;
+
+	temp = (size == 0) ? rb_get_min(tree) : rb_get_max(tree);
+	while (1) {
+		rx_node = z_rb_child(rx_node, size);
+		operations_count++;
+		if (rx_node == temp) {
+			break;
+		}
+	}
+
+	return operations_count;
+}
+
+int search_insert_remove_times(struct rbtree *tree)
+{
+	struct rbnode ins_nodes;
+	struct rbnode *ch = tree->root;
+	int side = 0, temp_count = 0;
+
+	rb_insert(tree, &ins_nodes);
+	/*operations(insert and remove) are same times in the same node*/
+	while (1) {
```

Comment:
General nitpick: a bug in the test logic or the tree will show up here as an infinite loop.  That **is** a test failure, but it's a very expensive one.  Though in practice this code and the test aren't actually buggy.

Reply:
I met this condition before. When I run the code, it will do as an infinite loop, until the stack broken. I think the reason is that inserting two same nodes into the tree, it will not work normally. So do you think how can I get the height of node in the tree instead this.:)

Reply:
I changed another method to implement that finding the height of the tree.